File: python_files/raylib/pong.py
import pyray as p

SCREEN_WIDTH: int = 1000
SCREEN_HEIGHT: int = 800
SCREEN_TITLE: str = "PONG"
SCREEN_BACKGROUND: p.Color = p.LIGHTGRAY


class Ball:
    def __init__(self) -> None:
        self.initial_x: float = p.get_screen_width() / 2
        self.initial_y: float = p.get_screen_height() / 2
        self.pos: p.Vector2 = p.Vector2(self.initial_x, self.initial_y)
        self.radius: float = 10.0
        self.initial_speed: float = 400.0
        self.speed: float = self.initial_speed
        self.speed_increment: float = 10.0
        self.direction: p.Vector2 = p.Vector2(
            -1 if p.get_random_value(0, 1) == 0 else -1,
            -1 if p.get_random_value(0, 1) == 0 else -1,
        )
        self.color: p.Color = p.RED
        self.active: bool = False

    def reset(self) -> None:
        self.pos = p.Vector2(self.initial_x, self.initial_y)
        self.direction = p.Vector2(
            -1 if p.get_random_value(0, 1) == 0 else -1,
            -1 if p.get_random_value(0, 1) == 0 else -1,
        )
        self.active = False
        self.speed = self.initial_speed

    def draw(self) -> None:
        p.draw_circle_v(self.pos, self.radius, self.color)

    def update(self) -> None:
        # activate/deactivate ball
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            self.active = not self.active

        # move ball
        if self.active:
            self.pos.x += self.direction.x * self.speed * p.get_frame_time()
            self.pos.y += self.direction.y * self.speed * p.get_frame_time()

        # normalize direction vector
        if self.direction.x != 0 and self.direction.y != 0:
            self.direction = p.vector2_normalize(self.direction)

        # ball bounds
        if (
            self.pos.y - self.radius < 0
            or self.pos.y + self.radius > p.get_screen_height()
        ):
            self.direction.y *= -1


class Paddle:
    def __init__(self, x: float) -> None:
        height: float = 100.0
        self.initial_y: float = p.get_screen_height() / 2 - height / 2
        self.rect: p.Rectangle = p.Rectangle(x, self.initial_y, 10.0, height)
        self.speed: float = 300.0
        self.color: p.Color = p.BLACK

    def reset(self) -> None:
        self.rect.y = self.initial_y

    def draw(self) -> None:
        p.draw_rectangle_rec(self.rect, self.color)

    def update_player(self) -> None:
        if p.is_key_down(p.KeyboardKey.KEY_UP) and self.rect.y > 0:
            self.rect.y -= self.speed * p.get_frame_time()
        if (
            p.is_key_down(p.KeyboardKey.KEY_DOWN)
            and self.rect.y < p.get_screen_height() - self.rect.height
        ):
            self.rect.y += self.speed * p.get_frame_time()

    def update_ai(self, ball: Ball) -> None:
        if ball.pos.y < self.rect.y + self.rect.height / 2 and self.rect.y > 0:
            self.rect.y -= self.speed * p.get_frame_time()
        if (
            ball.pos.y > self.rect.y + self.rect.height / 2
            and self.rect.y < p.get_screen_height() - self.rect.height
        ):
            self.rect.y += self.speed * p.get_frame_time()


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    score_left: int = 0
    score_right: int = 0

    ball: Ball = Ball()
    player: Paddle = Paddle(10.0)
    ai: Paddle = Paddle(p.get_screen_width() - player.rect.width - 10.0)

    while not p.window_should_close():
        ball.update()
        player.update_player()
        ai.update_ai(ball)

        # ball collision paddles
        if p.check_collision_circle_rec(ball.pos, ball.radius, player.rect):
            ball.direction.x *= -1
            ball.speed += ball.speed_increment
        if p.check_collision_circle_rec(ball.pos, ball.radius, ai.rect):
            ball.direction.x *= -1
            ball.speed += ball.speed_increment

        # update scores
        if ball.pos.x + ball.radius < 0:
            score_right += 1
            ball.reset()
            player.reset()
            ai.reset()
        if ball.pos.x - ball.radius > p.get_screen_width():
            score_left += 1
            ball.reset()
            player.reset()
            ai.reset()

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        # draw screen divider
        p.draw_line_ex(
            p.Vector2(p.get_screen_width() / 2, 0),
            p.Vector2(p.get_screen_width() / 2, p.get_screen_height()),
            5.0,
            p.GRAY,
        )

        # draw scores
        p.draw_text(str(score_left), 250, 30, 40, p.BLACK)
        p.draw_text(str(score_right), p.get_screen_width() - 250, 30, 40, p.BLACK)

        ball.draw()
        player.draw()
        ai.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: cpp_files/raylib/pong.cpp
#include "utils.h"
#include <raylib.h>
#include <raymath.h>

// BALL
typedef struct Ball
{
    Vector2 initial_pos;
    Vector2 pos;
    Vector2 direction;
    float radius;
    float initial_speed;
    float speed;
    float speed_increment;
    Color color;
    bool active;
    IntervalTimer interval_timer;

} Ball;

void
ball_init(Ball* ball)
{
    ball->initial_pos = (Vector2){ GetScreenWidth() * 0.5f, GetScreenHeight() * 0.5f };
    ball->pos = ball->initial_pos;
    ball->radius = 10.0f;
    ball->initial_speed = 300.0f;
    ball->speed = ball->initial_speed;
    ball->speed_increment = 10.0f;
    ball->direction.x = GetRandomValue(0, 1) == 0 ? -1 : 1;
    ball->direction.y = GetRandomValue(0, 1) == 0 ? -1 : 1;
    ball->color = RED;
    ball->active = false;

    interval_timer_init(&ball->interval_timer, 1.0f);
}

void
ball_reset(Ball* ball)
{
    ball->pos = ball->initial_pos;
    ball->speed = ball->initial_speed;
    ball->direction.x = GetRandomValue(0, 1) == 0 ? -1 : 1;
    ball->direction.y = GetRandomValue(0, 1) == 0 ? -1 : 1;
    ball->active = false;
}

void
ball_draw(Ball* ball)
{
    if (interval_timer_toggled(&ball->interval_timer) && !ball->active)
    {
        center_and_draw_text(
          "press SPACE to begin",
          30,
          (Rectangle){ 0, 0, ( float ) GetScreenWidth(), GetScreenHeight() + 300.0f });
    }

    DrawCircleV(ball->pos, ball->radius, ball->color);
}

void
ball_update(Ball* ball)
{
    interval_timer_update(&ball->interval_timer);

    // activate ball
    if (IsKeyPressed(KEY_SPACE))
    {
        ball->active = !ball->active;
    }

    // move ball
    if (ball->active)
    {
        ball->pos.x += ball->direction.x * ball->speed * GetFrameTime();
        ball->pos.y += ball->direction.y * ball->speed * GetFrameTime();
    }

    // normalize direction vector
    if (ball->direction.x != 0 && ball->direction.y != 0)
    {
        ball->direction = Vector2Normalize(ball->direction);
    }

    // ball bounds
    if (ball->pos.x < ball->radius || ball->pos.x > GetScreenWidth() - ball->radius)
    {
        ball->direction.x *= -1;
    }
    if (ball->pos.y < ball->radius || ball->pos.y > GetScreenHeight() - ball->radius)
    {
        ball->direction.y *= -1;
    }
}

// PADDLE
typedef struct Paddle
{
    Rectangle initial_rect;
    Rectangle rect;
    float speed;
    Color color;

} Paddle;

void
paddle_init(Paddle* paddle, float x)
{
    float height = 100.0f;
    paddle->initial_rect =
      (Rectangle){ x, ( float ) (GetScreenHeight()) / 2 - height / 2, 10.0f, height };
    paddle->rect = paddle->initial_rect;
    paddle->speed = 300.0f;
    paddle->color = BLACK;
}

void
paddle_draw(Paddle* paddle)
{
    DrawRectangleRec(paddle->rect, paddle->color);
}

void
paddle_update_player(Paddle* paddle)
{
    if (IsKeyDown(KEY_UP) && paddle->rect.y > 0)
    {
        paddle->rect.y -= paddle->speed * GetFrameTime();
    }
    if (IsKeyDown(KEY_DOWN) && paddle->rect.y < GetScreenHeight() - paddle->rect.height)
    {
        paddle->rect.y += paddle->speed * GetFrameTime();
    }
}

void
paddle_update_ai(Paddle* paddle, Ball* ball)
{
    if (ball->pos.y < paddle->rect.y + paddle->rect.height / 2 && paddle->rect.y > 0)
    {
        paddle->rect.y -= paddle->speed * GetFrameTime();
    }
    if (ball->pos.y > paddle->rect.y + paddle->rect.height / 2 &&
        paddle->rect.y < GetScreenHeight() - paddle->rect.height)
    {
        paddle->rect.y += paddle->speed * GetFrameTime();
    }
}

// MAIN
int
main(void)
{
    const int SCREEN_WIDTH = 800;
    const int SCREEN_HEIGHT = 600;
    const char SCREEN_TITLE[] = "Pong";
    const Color SCREEN_BACKGROUND = LIGHTGRAY;

    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

    int score_left = 0;
    int score_right = 0;

    Ball ball;
    Paddle player;
    Paddle ai;

    ball_init(&ball);
    paddle_init(&player, 10.0f);
    paddle_init(&ai, GetScreenWidth() - player.rect.width - 10.0f);

    while (!WindowShouldClose())
    {
        ball_update(&ball);
        paddle_update_player(&player);
        paddle_update_ai(&ai, &ball);

        BeginDrawing();
        ClearBackground(SCREEN_BACKGROUND);

        center_and_draw_text(TextFormat("%d", score_left),
                             40,
                             (Rectangle){ 0, 0, GetScreenWidth() * 0.5f, 100.0f },
                             BLACK);
        center_and_draw_text(
          TextFormat("%d", score_left),
          40,
          (Rectangle){ GetScreenWidth() * 0.5f, 0, GetScreenWidth() * 0.5f, 100.0f },
          BLACK);

        ball_draw(&ball);
        paddle_draw(&player);
        paddle_draw(&ai);

        EndDrawing();
    }

    CloseWindow();

    return 0;
}